internal/service/kafka/resource_kafka_topic_test.go

```diff
@@ -341,3 +345,144 @@ func TestPartitions(t *testing.T) {
 		})
 	}
 }
+
+// TestAccAivenKafkaTopic_recreate validates that topic is recreated if it is missing
+// Kafka looses all topics on turn off/on, then TF recreates topics. This test imitates the case.
+func TestAccAivenKafkaTopic_recreate_missing(t *testing.T) {
+	project := os.Getenv("AIVEN_PROJECT_NAME")
+
+	prefix := "test-tf-acc-" + acctest.RandString(7)
+	kafkaResource := "aiven_kafka.kafka"
+	topicResource := "aiven_kafka_topic.topic"
+	kafkaName := prefix + "-kafka"
+	topicName := "topic"
+	kafkaID := fmt.Sprintf("%s/%s", project, kafkaName)
+	topicID := kafkaID + "/topic"
+
+	config := testAccAivenKafkaTopicResourceRecreateMissing(prefix, project)
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:          func() { acc.TestAccPreCheck(t) },
+		ProviderFactories: acc.TestAccProviderFactories,
+		CheckDestroy:      testAccCheckAivenKafkaTopicResourceDestroy,
+		Steps: []resource.TestStep{
+			{
+				// Step 1: setups resources, creates the state
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(kafkaResource, "id", kafkaID),
+					resource.TestCheckResourceAttr(topicResource, "id", topicID),
+				),
+			},
+			{
+				// Step 2: deletes topic, then runs apply, same config & checks
+				PreConfig: func() {
+					client, err := provider.DefaultAivenClient()
```

Comment:
an instance of `*aiven.Client` is already available by calling `provider.Provider().Meta().(*aiven.Client)`, so I'm not sure if this is justified to have a separate exported constructor function for the client

moreover, the constructor function appears to violate the `provider` package's semantic, and even if we do want to have it this way, it should probably be moved to some other package that makes more sense for it, e.g. a new package called `client` or `clientutil`, although, it probably worths to address the aforementioned necessity of obtaining the client in this manner, first

Reply:
moreover, the constructor function appears to violate the `provider` package's semantics, and even if we do want to have it this way, it should probably be moved to some other package that makes more sense for it, e.g. a new package called `client` or `clientutil`, although, it probably worths to address the aforementioned necessity of obtaining the client in this manner, first

Reply:
`provider.Provider().Meta().(*aiven.Client)` 
Doesn't work. It doesn't cast to the client.  
Moved client constructors to a separate package.

Reply:
Found it `acc.TestAccProvider.Meta().(*aiven.Client)`

Reply:
> Moved client constructors to a separate package.

I don't quite understand why do we need to have them at all, since we shouldn't be constructing multiple clients in the provider anyway

Reply:
But we have them in provider, sweep and example tests.

Reply:
did you forget to update the example tests?

Reply:
No. It's [there](https://github.com/aiven/terraform-provider-aiven/pull/1008/files#diff-25fd8eb50f1be757367ab21a98d4ee45ac5eae017f561fa4fa72a2f41430a4cbR91).